### DISCUSSION TITLE
Custom module is not registering

### DISCUSSION THREAD

**Discussion by antick** [directus/directus#9852] [Archive]
Hi,

I am currently using `9.0.0-rc.96`.

I have gone through the following guide for setting up a custom module:

https://docs.directus.io/extensions/modules/

I created a very simple vue 3 app with this content in a directory outside the directus:

```js
<template>
  <private-view title="My Custom Module">Content goes here...</private-view>
</template>

<script>
export default {};
</script>
```

and in the above vue3 directory I added the following `index.js`:

```js
import ModuleComponent from './App.vue';

export default {
  id: 'importer',
  name: 'Importer',
  icon: 'box',
  routes: [
    {
      path: '',
      component: ModuleComponent,
    },
  ],
};
```

Then I ran `npx directus-extension build` and it generated a `dist/index.js` file. So I moved it in my directus setup in the following path:

`<directus>/extensions/modules/importer/index.js`

Now I ran the `npm run build` and `npm run dev` in directus but was still not able to see this custom module in the Project settings list.

I also looked into the directus code. I think this is the file that registers the custom modules and merges them with the in-built modules.

`app/src/modules/register.ts`

I tried to print the `customModules` in the register file for this code:

```js
const customModules: { default: ModuleConfig[] } = import.meta.env.DEV
			? await import('@directus-extensions-module')
			: await import(/* @vite-ignore */ `${getRootPath()}extensions/modules/index.js`);

		modules.push(...customModules.default);
```

But the `customModules` was empty. And in my case, it is going to the true block.

What could be the reason that it's not loading the custom component at all? And looking at the code above in the `register.ts`, I am not sure how does it actually work for the multiple custom modules. I checked the 9.0.0-rc.80's register.ts code where we were looping through all the extensions: (https://github.com/directus/directus/blob/v9.0.0-rc.80/app/src/modules/register.ts) which made sense.

And I am also not able to understand how this `${getRootPath()}extensions/modules/index.js` will work when we are never putting our generated files in the modules folder directly. We always create our custom module folder first and then we put the index file there.

Can anyone please help me understand how can I get these custom modules working?

**Comment by rijkvanzanten** [comment]
We've vastly overhauled how extensions are bundled and registered for the final production release (9.0.0). The current docs are up to date and confirmed working 🙂

**Reply by antick** [reply]
Thanks. I got this working. I was running `npm run dev` to start the project and it wasn't picking up the extensions. Once I started the project by executing `npx directus start`, my module started showing.

It was a confusion on my end between `npm run dev` and `npx directus start`.

**Reply by nickrum** [reply]
`npm run dev` starts Directus in dev mode which should only be used when actually working on the Directus code base.

**Comment by ad2ien** [comment]
Hi, I have the same issue with version 9.4.0. 
- In a separate folder :
  - `npm init directus-extension`
  - `npm run build`
  - copy the content of dist folder in extensions/modules/my-module
- `npx directus start`
-> there's a `Loaded extensions: my-module` trace but noting new shows in the module's list

And it's working with a panel extension.
Thx.

**Reply by rijkvanzanten** [reply]
Did you enable your custom module in project settings?

**Reply by ad2ien** [reply]
Oooook indeed I had to activate it. Thanks!

**Comment by nickrum** [comment]
> Then I ran `npx directus-extension build` and it generated a `dist/index.js` file. So I moved it in my directus setup in the following path:
> 
> `<directus>/extensions/modules/importer/index.js`

When running Directus in dev mode, the App is responsible for loading App extensions. They are loaded from `<directus-repo>/api/extensions/`.

> And looking at the code above in the `register.ts`, I am not sure how does it actually work for the multiple custom modules.

The API server bundles all extensions of a certain type to a single file to reduce the amount of requests to the server.

> And I am also not able to understand how this `${getRootPath()}extensions/modules/index.js` will work when we are never putting our generated files in the modules folder directly.

This `${getRootPath()}extensions/modules/index.js` is only a virtual path that doesn't correlate with the file system. After bundling the extensions, the API server will server the extension source under this URL.

**Reply by antick** [reply]
Thanks for the explanation. Now it all makes sense.
